Stop using a CMBufferQueueRef for re-ordering samples.

jyavenard opened the pull request:
#### c755d07d28c8ebe6f680c7bd19045baf8c0f31ce
<pre>
Stop using a CMBufferQueueRef for re-ordering samples.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295787">https://bugs.webkit.org/show_bug.cgi?id=295787</a>
<a href="https://rdar.apple.com/155621751">rdar://155621751</a>

Reviewed by Gerald Squelart.

CMBufferQueue is a priority queue using a CFArray as backend.
It&apos;s not particularly adapted for our use as a dequeue where we will append
on one end to pop on the other.
It&apos;s also particularly inefficient in adding samples that aren&apos;t out of order.

So we introduced a MediaReorderQueue that is using a Deque as backend
and will only perform re-ordering as needed. So for YouTube in particular
adding and removing a frame are in O(1).
When re-ordering the frame, we will use bubble sort from the end of the queue
which is very optimal for our use case where the queue is already ordered
and only a few samples may be out of order at the end.

We migrate most of VideoMediaSampleRenderer use of CoreMedia objects
to webkit&apos;s MediaTime/MediaSample for ease.

YouTube Profiling shows a significant reduction in memory allocations and
time spent in VideoMediaSampleRenderer::purgeDecodedSampleQueue.

Adding API tests, no observable changes, covered by existing tests.

* Source/WebCore/platform/graphics/MediaReorderQueue.h: Added.
(WTF::CFTypeTrait&lt;CMSampleBufferRef&gt;::typeID): Deleted.
(WebCore::getDecodeTime): Deleted.
(WebCore::getPresentationTime): Deleted.
(WebCore::getDuration): Deleted.
(WebCore::compareBuffers): Deleted.
(WebCore::createBufferQueue): Deleted.
(WebCore::sampleCallback): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/MediaReorderQueue.cpp: Added.
(TestWebKitAPI::TestSample::create):
(TestWebKitAPI::TestSample::TestSample):
(TestWebKitAPI::TEST(MediaReorderQueue, MediaSample)):
(TestWebKitAPI::TEST(MediaReorderQueue, Int)):
(TestWebKitAPI::TEST(MediaReorderQueue, Pair)):

Canonical link: <a href="https://commits.webkit.org/297392@main">https://commits.webkit.org/297392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/392ce3cfc0ba09ab16632c321ec40f97775f6373

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117675 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113602 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39892 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84819 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65260 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24875 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18613 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61507 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120923 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38693 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28746 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93719 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93545 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23845 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38684 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16477 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34729 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44067 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38245 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41581 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39947 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/77ac849f818ff568e8ebd4f3e52dafa3352a64b3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47894

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/47d5729e37b258fa12f0183e17d431fb14e64e42. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47894

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f0c2df5094694fada7afa34d6eb54861520fdfb3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47894

squelart reviewed (approved):
r+ mostly for the new `MediaReorderQueue`.
I'd suggest getting a review from someone in media who can verify all the other changes, some are not as trivial as just changing a type or function name!

jyavenard commented on Source/WebCore/platform/graphics/MediaReorderQueue.h:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    void clear() { m_queue.clear(); }
+    uint8_t reorderSize() const { return m_reorderSize; }
+    void setReorderSize(uint8_t size) { m_reorderSize = size; }
+    bool isAvailable() const { return m_queue.size() > m_reorderSize; }
```
They will be required to replace the H265 reorder queue found in webrtc and that we want to move to WebCore

squelart commented on Source/WebCore/platform/graphics/MediaReorderQueue.h:
```diff
@@ -0,0 +1,124 @@
+
+    std::optional<T> takeIfAvailable(bool& moreObjectsAvailable)
+    {
+        moreObjectsAvailable = false;
```
Sorry if I was unclear, I was suggesting:
```
        if (!isAvailable()()) {
            moreObjectsAvailable = false;
            return { };
        }
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bfe7793ba7d3d567682155226fd636fcbe557a73. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47894

jyavenard commented on Tools/TestWebKitAPI/Tests/WebCore/MediaReorderQueue.cpp:
```diff
@@ -0,0 +1,173 @@
+    }
+}
+
+} // namespace TestWebKitAPI
```
added tests

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/392ce3cfc0ba09ab16632c321ec40f97775f6373. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47894

jyavenard committed:
Stop using a CMBufferQueueRef for re-ordering samples.
https://bugs.webkit.org/show_bug.cgi?id=295787
rdar://155621751

Reviewed by Gerald Squelart.

CMBufferQueue is a priority queue using a CFArray as backend.
It's not particularly adapted for our use as a dequeue where we will append
on one end to pop on the other.
It's also particularly inefficient in adding samples that aren't out of order.

So we introduced a MediaReorderQueue that is using a Deque as backend
and will only perform re-ordering as needed. So for YouTube in particular
adding and removing a frame are in O(1).
When re-ordering the frame, we will use bubble sort from the end of the queue
which is very optimal for our use case where the queue is already ordered
and only a few samples may be out of order at the end.

We migrate most of VideoMediaSampleRenderer use of CoreMedia objects
to webkit's MediaTime/MediaSample for ease.

YouTube Profiling shows a significant reduction in memory allocations and
time spent in VideoMediaSampleRenderer::purgeDecodedSampleQueue.

Adding API tests, no observable changes, covered by existing tests.

* Source/WebCore/platform/graphics/MediaReorderQueue.h: Added.
(WTF::CFTypeTrait<CMSampleBufferRef>::typeID): Deleted.
(WebCore::getDecodeTime): Deleted.
(WebCore::getPresentationTime): Deleted.
(WebCore::getDuration): Deleted.
(WebCore::compareBuffers): Deleted.
(WebCore::createBufferQueue): Deleted.
(WebCore::sampleCallback): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/MediaReorderQueue.cpp: Added.
(TestWebKitAPI::TestSample::create):
(TestWebKitAPI::TestSample::TestSample):
(TestWebKitAPI::TEST(MediaReorderQueue, MediaSample)):
(TestWebKitAPI::TEST(MediaReorderQueue, Int)):
(TestWebKitAPI::TEST(MediaReorderQueue, Pair)):

Canonical link: https://commits.webkit.org/297392@main

webkit-commit-queue commented:
Committed 297392@main (c755d07d28c8): <https://commits.webkit.org/297392@main>

Reviewed commits have been landed. Closing PR #47894 and removing active labels.
